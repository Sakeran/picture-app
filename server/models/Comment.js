const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const moment = require('moment');

const commentSchema = new Schema({
  user: {
    type: Schema.Types.ObjectId,
    ref: 'User',
  },
  post: {
    type: Schema.Types.ObjectId,
    ref: 'Post'
  },
  text: String,
  deleted: {
    type: Boolean,
    default: false
  },
  deletedText: {
    type: String,
    get: (string) => {
      // This should only be viewable from the database itself.
      return '';
    }
  },
  createdAt: {
    type: Date,
    default: Date.now,
  },
});

commentSchema.virtual('date').get(function() {
  return moment(this.createdAt).format('MMMM Do YYYY');
});

// Deletes the comment if the user created the post or is an admin.
// Should begin displaying a default message instead of the actual text.
commentSchema.methods.deleteIfUserAllowed = function(user) {
  return new Promise((resolve, reject) => {
    if (!user || user.constructor.modelName !== 'User') {
      return reject(new Error('User must be a User model'));
    }
    const isAdmin = user.hasAdmin;
    const isOwner = this.user.equals(user.id);
    if (!isOwner && !isAdmin) {
      return resolve(false);
    }
    this.deleted = true;
    this.deletedText = this.text;
    this.text = '(This comment has been removed.)';
    this.save()
    .then(() => resolve(true));
  });
};

const Comment = mongoose.model('Comment', commentSchema);

module.exports = Comment;
