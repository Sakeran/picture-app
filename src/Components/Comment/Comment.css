.Comment {
  margin: 0.5em 0;
  padding: 0.5em 0;
  background: white;
}

.Comment-header {
  margin: 0;
  padding-left: 0.5em;
  padding-bottom: 0.25em;
  font-weight: 400;
  border-bottom: 1px solid #6d1cba;
}

.Comment-header a {
  color: #6d1cba;
}

.Comment-text {
  padding-left: 0.5em;
}
