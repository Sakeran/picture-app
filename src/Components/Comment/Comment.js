import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

import './Comment.css';

// Determines if the delete option should be shown for a given comment.
const showDeleteOption = (commentId, deleted, creator, user, deleteFn) => {
  if (deleted || !user || !deleteFn) { return false; }
  if (user.isAdmin || (user.id === creator.id) ) { return true; }
  return false;
};

const Comment = ({comment: {user: creator, date, text, deleted, id},
                  user,
                  deleteFn}) => (
  <div className="Comment">
    <h4 className="Comment-header">
      <Link to={`/profile/${creator.id}`}>{creator.username}</Link> commented on {date}
      {showDeleteOption(id, deleted, creator, user, deleteFn) &&
        <button onClick={deleteFn}>(Delete This Comment)</button>
      }
    </h4>
    <p className="Comment-text">{text}</p>
  </div>
);

Comment.propTypes = {
  comment: PropTypes.object.isRequired,
  user: PropTypes.object,
  deleteFn: PropTypes.func,
};

export default Comment;
