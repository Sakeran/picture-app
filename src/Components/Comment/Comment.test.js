import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Comment from './Comment';

const testComment = {
  text: 'Test Comment',
  user: {
    username: 'Test User'
  },
  date: 'January 1st 2001'
};

test('Comment component renders correctly', () => {
  const component = shallow(<Comment comment={testComment} />);
  expect(component.text()).toContain('Test Comment');
  expect(component.text()).toContain('Test User');
  expect(component.text()).toContain('January 1st 2001');
  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
