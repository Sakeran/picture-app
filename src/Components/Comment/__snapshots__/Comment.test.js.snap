// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment component renders correctly 1`] = `
<div
  className="Comment"
>
  <h4
    className="Comment-header"
  >
    <Link
      replace={false}
      to="/profile/1111"
    >
      Test User
    </Link>
     commented on 
    January 1st 2001
  </h4>
  <p
    className="Comment-text"
  >
    Test Comment
  </p>
</div>
`;
